### PR TITLE
Heedls 681 Add LH SSO Account Linking Endpoint

### PR DESCRIPTION
**Pull request by wilkinsm**
### JIRA link
[https://softwiretech.atlassian.net/browse/HEEDLS-681](https://softwiretech.atlassian.net/browse/HEEDLS-681)

### Description
Add new endpoint that recieves requests from the LH that links the LH accounts to the DLS accounts.
This is part of the overall Learning Hub SSO linking journey.

### Screenshots
**Desktop standard**
image

**Desktop Already Linked**
image

**Desktop Invalid resource id**
image

**Tablet**
image

**Mobile**
image

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

**Comment by wilkinsm** [comment]
Need to write up about the new Integration Tests!

File: DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs [hunk 1]
```diff
@@ -154,6 +169,134 @@ public void VerifyHash_returns_false_for_hashes_created_outside_tolerance_time([
...
+        }
+
+        [Test]
+        public void ParseSsoAccountLinkingRequest_throws_exception_when_state_is_incomplete()
+        {
+            // Given
+            var linkLearningHubRequest = new LinkLearningHubRequest
+            {
+                Hash = "vrdz",
+                SessionIdentifier = Guid.NewGuid(),
```
Comment (JBsoftwire) [inline DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs:229]: If I understand this correctly, this setup would still fail if the state was complete because the two sessionIdentifiers will differ. I think it would be better if each test request was as valid as possible save for the one error. This may also apply to the `verifyHash_returns_false` test.

**Review by JBsoftwire — CHANGES_REQUESTED** [review]
Just a couple of things I noticed about the tests.

File: DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs [hunk 2]
```diff
@@ -948,6 +948,21 @@ public void GetSupervisorsAtCentreForCategory_returns_expected_admins()
             result.Should().OnlyContain(au => au.CategoryId == 0 || au.CategoryId == 1);
         }
 
+        [Test]
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Data.Tests/Services/UserServiceTests.cs:951]: Ah yes good spot. Done.

File: DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs (hunk 1 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs]: Although both of these tests do "fail" for the right reason I get your point so I have updated them ensuring the input data represents more clearly what is being tested for.

**Review by JBsoftwire — APPROVED** [review]
Looks good.

File: DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    using DigitalLearningSolutions.Data.Services;
+    using DigitalLearningSolutions.Web.Helpers;
+    using DigitalLearningSolutions.Web.ViewModels.Signposting.LinkLearningHubSso;
+    using Microsoft.AspNetCore.Authorization;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.FeatureManagement.Mvc;
+
+    [FeatureGate(FeatureFlags.UseSignposting)]
+    [Route("Signposting/[Controller]")]
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs:15]: Would there be any issue with making this route `Signposting/LinkLearningHubSso`?

File: DigitalLearningSolutions.Web/Views/Signposting/LinkLearningHubSso/Index.cshtml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+<div class="nhsuk-grid-row">
+  <div class="nhsuk-grid-column-full">
+
+    <h1 class="nhsuk-heading-xl">Account linked</h1>
+
+    @if (Model.ShowIsAlreadyLinkedWarning)
+    {
+      <p>Your account has already been linked with the Learning Hub in a previous link request.</p>
+      <p>Therefore no changes have been made to your account regarding this link request.</p>
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web/Views/Signposting/LinkLearningHubSso/Index.cshtml:22]: I think we could make this a bit more concise (Eg. `Your account is already linked with Learning Hub. \n No change has been made.`), but let's check with HEE when we demo it.

File: DigitalLearningSolutions.Web/ViewModels/Signposting/LinkLearningHubSso/LinkLearningHubViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+            ShowIsAlreadyLinkedWarning = isAccountAlreadyLinked;
+            ResourceLinkId = learningHubResourceId;
+            ShowResourceLink = learningHubResourceId.HasValue;
+        }
+
+        public bool ShowIsAlreadyLinkedWarning { get; set; }
+
+        public int? ResourceLinkId { get; set; }
+
+        public bool ShowResourceLink { get; set; }
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web/ViewModels/Signposting/LinkLearningHubSso/LinkLearningHubViewModel.cs:16]: This could be slightly clearer as `=> ResourceLinkId.HasValue`

File: DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs [hunk 6]
```diff
@@ -154,6 +171,139 @@ public void VerifyHash_returns_false_for_hashes_created_outside_tolerance_time([
...
 
+        [Test]
+        public void ParseSsoAccountLinkingRequest_throws_exception_when_verifyHash_returns_false()
+        {
+            // Given
+            var linkLearningHubRequest = new LinkLearningHubRequest
+            {
+                Hash = "vrdz",
+                UserId = 56789,
+            };
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs:182]: We should make the session identifier non-null so we know that's not causing the exception.

File: DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs [hunk 7]
```diff
@@ -154,6 +171,139 @@ public void VerifyHash_returns_false_for_hashes_created_outside_tolerance_time([
...
+        }
+
+        [Test]
+        public void ParseSsoAccountLinkingRequest_throws_exception_when_state_sessionIdentifier_is_not_guid()
+        {
+            // Given
+            var linkLearningHubRequest = new LinkLearningHubRequest
+            {
+                Hash = "vrdz",
+                SessionIdentifier = Guid.NewGuid(),
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs:255]: Shouldn't this also be `no-a-guid`?

File: DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserDataServiceTests.cs [hunk 8]
```diff
@@ -418,5 +418,57 @@ public void ActivateDelegateUser_activates_delegate_user()
...
+
+            // Then
+            var result = userDataService.GetDelegateUserLearningHubAuthId(delegateId);
+
+            result.Should().NotBeNull()
+                .And.Subject.Should().Be(learningHubAuthId);
+        }
+
+        [Test]
+        public void GetDelegateUserLearningHubAuthId_returns_expected_delegate_learningHubAuthId()
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserDataServiceTests.cs:442]: This is the same as the test above but with different values right?

**Review by SteveJacksonSoft — CHANGES_REQUESTED** [review]
Looked at everything except integration tests so far.

File: DigitalLearningSolutions.Web.Tests/ControllerHelpers/ControllerContextHelper.cs [hunk 9]
```diff
@@ -113,6 +116,20 @@ string cookieValue
             return controller;
         }
 
+        public static T WithMockSessionData<T>(this T controller, IDictionary<string, string> sessionData)
+            where T : Controller
+        {
+            var fakeSession = A.Fake<ISession>();
+            controller.HttpContext.Session = fakeSession;
+            byte[]? outValue;
+            A.CallTo(() => fakeSession.TryGetValue(A<string>._, out outValue))
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web.Tests/ControllerHelpers/ControllerContextHelper.cs:125]: This would result in an error of `expression tree may not contain a discard`

File: DigitalLearningSolutions.Web/Styles/layout.scss [hunk 10]
```diff
@@ -70,6 +70,17 @@ body:not(.js-enabled) {
   }
 }
 
+.nhsuk-header--transactional {
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Styles/layout.scss:73]: `nhsuk-header--transactional` is part of the NHS css package - these are just overrides to cover that specific scenario.

File: DigitalLearningSolutions.Web.IntegrationTests/xunit.runner.json [hunk 11]
```diff
@@ -1,3 +1,4 @@
 {
-  "shadowCopy": false
+  "shadowCopy": false,
+  "parallelizeTestCollections":  false
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web.IntegrationTests/xunit.runner.json:3]: Yes, given the new integration tests, they were being run in parallel with the current ones causing a conflict with `StartUp` and registered services.

File: DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml [hunk 12]
```diff
@@ -54,7 +54,7 @@
   <script>document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');</script>
 
   <a class="nhsuk-skip-link" href="#maincontent">Skip to main content</a>
-  <header class="nhsuk-header nhsuk-header--organisation nhsuk-header--white" role="banner">
+  <header class="nhsuk-header @((bool?)ViewData[LayoutViewDataKeys.TransactionalHeader] == true ? "nhsuk-header--transactional" : "nhsuk-header--organisation nhsuk-header--white")" role="banner">
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml:57]: As per the bellow screenshot. 
You get a white header instead of the blue as per the wireframe.
In this screenshot there is no logo present (I've marked a yellow blob where it normally is).
In the account I used when I created this a logo did appear there but in the account I've logged on as today there is no logo.
Maybe the logo change is not required. What do you think?

[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml:57]: image

File: DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs (hunk 3 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs]: No issue. Done.

File: DigitalLearningSolutions.Data/Services/LearningHubSsoSecurityService.cs [hunk 13]
```diff
@@ -56,6 +61,60 @@ public bool VerifyHash(string state, string hash)
...
+
+        private void ValidateUserState(Guid receivedSessionIdentifier, Guid? storedSessionIdentifier)
+        {
+            if (!storedSessionIdentifier.HasValue || receivedSessionIdentifier != storedSessionIdentifier.Value)
+            {
+                throw new LearningHubLinkingRequestException("Invalid Learning Hub linking session.");
+            }
+        }
+
+        private (Guid sessionIdentifier, int? resourcesId) ParseState(LinkLearningHubRequest linkLearningHubRequest)
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Data/Services/LearningHubSsoSecurityService.cs]: Done.

File: DigitalLearningSolutions.Data/Services/LearningHubSsoSecurityService.cs [hunk 14]
```diff
@@ -56,6 +61,60 @@ public bool VerifyHash(string state, string hash)
...
+        {
+            var stateItems = linkLearningHubRequest.State.Split("_refId:");
+
+            if (stateItems.Length != 2)
+            {
+                throw new LearningHubLinkingRequestException("Invalid Learning Hub linking state.");
+            }
+
+            if (!Guid.TryParse(stateItems.ElementAt(0), out var validSessionIdentifier))
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Data/Services/LearningHubSsoSecurityService.cs]: No fundamental difference here. I'm more used to writing `ElementAt` but have switched to `[]` as I see that is what is used elsewhere.

File: DigitalLearningSolutions.Data/Services/LearningHubSsoSecurityService.cs [hunk 15]
```diff
@@ -56,6 +61,60 @@ public bool VerifyHash(string state, string hash)
...
+        private void ValidateUserIdFromHash(LinkLearningHubRequest linkLearningHubRequest)
+        {
+            var isVerified = VerifyHash(
+                linkLearningHubRequest.UserId.ToString(),
+                linkLearningHubRequest.Hash
+            );
+
+            if (!isVerified)
+            {
+                throw new LearningHubLinkingRequestException("Invalid Learning Hub UserId.");
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Data/Services/LearningHubSsoSecurityService.cs]: Fair point. Done.

File: DigitalLearningSolutions.Data/Services/LearningHubSsoSecurityService.cs [hunk 16]
```diff
@@ -1,16 +1,21 @@
...
+    using DigitalLearningSolutions.Data.Models.Signposting.LinkLearningHubSso;
     using Microsoft.Extensions.Configuration;
 
     public interface ILearningHubSsoSecurityService
     {
         string GenerateHash(string state);
 
         bool VerifyHash(string state, string hash);
+
+        int? ParseSsoAccountLinkingRequest(LinkLearningHubRequest linkLearningHubRequest);
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Data/Services/LearningHubSsoSecurityService.cs]: Nice. There were originally separate.
I joined them given the association and so I could more easily use the `VerifyHash` and simplify the setup.
I've split them back out again now.

File: DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+            if (!ModelState.IsValid)
+            {
+                throw new LearningHubLinkingRequestException("Invalid Learning Hub request.");
+            }
+
+            linkLearningHubRequest.SessionIdentifier = Guid.Parse(
+                HttpContext.Session.GetString(LinkLearningHubRequest.SessionIdentifierKey)
+            );
+            var learningHubResourcedId =
+                learningHubSsoSecurityService.ParseSsoAccountLinkingRequest(linkLearningHubRequest);
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs]: I know what you mean. The `Parse..` bit is ok but yes this method is overloaded which more than that.
I've renamed.

File: DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    using DigitalLearningSolutions.Web.ViewModels.Signposting.LinkLearningHubSso;
+    using Microsoft.AspNetCore.Authorization;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.FeatureManagement.Mvc;
+
+    [FeatureGate(FeatureFlags.UseSignposting)]
+    [Route("Signposting/[Controller]")]
+    [Authorize(Policy = CustomPolicies.UserOnly)]
+    public class LinkLearningHubSsoController : Controller
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs]: Yeah thats a fair point. Done.

File: DigitalLearningSolutions.Web/Views/Signposting/LinkLearningHubSso/Index.cshtml (hunk 4 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Views/Signposting/LinkLearningHubSso/Index.cshtml]: I've changed it as per the suggestion but we can still bring this up.

File: DigitalLearningSolutions.Web/ViewModels/Signposting/LinkLearningHubSso/LinkLearningHubViewModel.cs (hunk 5 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/ViewModels/Signposting/LinkLearningHubSso/LinkLearningHubViewModel.cs]: Done.

File: DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs (hunk 6 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs]: I've removed that property as it's technically not part of the "incoming" request.

File: DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs (hunk 7 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Data.Tests/Services/LearningHubSsoSecurityServiceTests.cs]: This property has been removed.

File: DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserDataServiceTests.cs (hunk 8 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserDataServiceTests.cs]: Yes it is. Nice. This one was added during markups and although the SIT is different it is essentially duplication.

File: DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+        public LinkLearningHubSsoController(
+            ILearningHubSsoSecurityService learningHubSsoSecurityService,
+            IUserService userService
+        )
+        {
+            this.learningHubSsoSecurityService = learningHubSsoSecurityService;
+            this.userService = userService;
+        }
+
+        public IActionResult Index([FromQuery] LinkLearningHubRequest linkLearningHubRequest)
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs]: Any exceptions thrown will be automatically logged by Seri.

File: DigitalLearningSolutions.Data.Tests/Services/LearningHubLinkServiceTests.cs [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+        private ILearningHubSsoSecurityService learningHubSsoSecurityService = null!;
+        private IUserDataService userDataService = null!;
+
+        [SetUp]
+        public void Setup()
+        {
+            userDataService = A.Fake<IUserDataService>();
+            learningHubSsoSecurityService = A.Fake<ILearningHubSsoSecurityService>();
+            A.CallTo(() => learningHubSsoSecurityService.VerifyHash("56789", A<string>._)).Returns(false);
+            A.CallTo(() => learningHubSsoSecurityService.VerifyHash("12345", A<string>._)).Returns(true);
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Tests/Services/LearningHubLinkServiceTests.cs:24]: Might be nice to change these to "invalid-hash" and "valid-hash" respectively to make the contents of the tests a bit more understandable, or just set it up in each test.

File: DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserDataServiceTests.cs [hunk 21]
```diff
@@ -438,31 +438,13 @@ public void SetDelegateUserLearningHubAuthId_correctly_sets_delegates_learningHu
...
-        }
-
         [Test]
         public void GetDelegateUserLearningHubAuthId_returns_null_delegate_learningHubAuthId()
         {
             using var transaction = new TransactionScope();
 
             // Given
-            const int delegateId = 11;
+            const int delegateId = 3;
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserDataServiceTests.cs:447]: Why the change?

File: DigitalLearningSolutions.Web.Tests/ControllerHelpers/ControllerContextHelper.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web.Tests/ControllerHelpers/ControllerContextHelper.cs:125]: Interesting. Fair.

File: DigitalLearningSolutions.Web/Styles/layout.scss (hunk 10 above)
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web/Styles/layout.scss:73]: Grand.

File: DigitalLearningSolutions.Web.IntegrationTests/TestHelpers/TestUserApp.cs [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        public Action<IApplicationBuilder> Configure(Action<IApplicationBuilder> next)
+        {
+            return builder =>
+            {
+                next(builder);
+
+                builder.MapWhen(
+                    context => context.Request.Path.Value.StartsWith("/SetDelegateTestSession"),
+                    loginApp =>
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web.IntegrationTests/TestHelpers/TestUserApp.cs:22]: To check my understanding, this is effectively adding a new endpoint at `/SetDelegateTestSession` right?

File: DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        protected sealed override IWebHostBuilder CreateWebHostBuilder()
+        {
+            return base.CreateWebHostBuilder();
+        }
+
+        protected override void ConfigureWebHost(IWebHostBuilder builder)
+        {
+            builder
+                .UseEnvironment("SIT")
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs:22]: Does this do anything special? Would it change behaviour if this string were changed?

File: DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+                .ConfigureServices(
+                    services =>
+                    {
+                        services.Configure<MvcOptions>(
+                            o => { o.Filters.Remove(new AutoValidateAntiforgeryTokenAttribute()); }
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs:27]: This filter forces anti forgery validation to be used.
In order to validate Anitforgery tokens it requires a bunch of other setup which I think just over complicates the tests and also is something I don't think we need to do given it's generally not part of the scope of the tests.

File: DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs (hunk 23 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs:22]: Yes, this will ensure the right appSettings file is used.

File: DigitalLearningSolutions.Web.IntegrationTests/TestHelpers/TestUserApp.cs (hunk 22 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Web.IntegrationTests/TestHelpers/TestUserApp.cs:22]: Yes it does, it sets the endpoint and acts like an auto login.

File: DigitalLearningSolutions.Web.IntegrationTests/CustomWebApplicationFactory.cs [hunk 25]
```diff
@@ -1,30 +0,0 @@
...
-            builder.ConfigureServices(services =>
-            {
-                var defaultSqlConnection = services.SingleOrDefault(
-                    service => service.ServiceType == typeof(IDbConnection));
-
-                services.Remove(defaultSqlConnection);
-
-                var config = ConfigHelper.GetAppConfig();
-                var connectionString = config.GetConnectionString(ConfigHelper.UnitTestConnectionStringName);
-                services.AddScoped<IDbConnection>(_ => new SqlConnection(connectionString));
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web.IntegrationTests/CustomWebApplicationFactory.cs:26]: The integration tests are now setup to ensure they look at the SIT appSettings file.
Given that, this code is just duplicate of what is happening in the main StartUp class.
We do still have a DB connection for integration tests.

File: DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserDataServiceTests.cs (hunk 21 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserDataServiceTests.cs:447]: There was no reason to target different users for the two tests. Either should produce the same result.

File: DigitalLearningSolutions.Web.IntegrationTests/TestHelpers/TestUserApp.cs [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+    using System;
+    using System.Linq;
+    using System.Security.Claims;
+    using DigitalLearningSolutions.Web.Helpers;
+    using Microsoft.AspNetCore.Authentication;
+    using Microsoft.AspNetCore.Builder;
+    using Microsoft.AspNetCore.Hosting;
+    using Microsoft.AspNetCore.Http;
+
+    public class TestUserAppStartupFilter : IStartupFilter
```
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web.IntegrationTests/TestHelpers/TestUserApp.cs]: Done.

File: DigitalLearningSolutions.Data.Tests/Services/LearningHubLinkServiceTests.cs (hunk 20 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Data.Tests/Services/LearningHubLinkServiceTests.cs]: Yeah makes sense, done.

File: DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml (hunk 12 above)
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml:57]: So the aim here is that the nav bar will remain consistent in a transition from the recommended learning page. That in turn is consistent with the nav bar in a launched course / self assessment. I think that may depend on the centre you're in and whether it has a logo. Do you know more? Otherwise I plan to look through that later on today hopefully.

File: DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs:22]: Cool, can we just have a comment here saying that we're removing this to avoid unnecessary setup?

File: DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs (hunk 23 above)
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs:17]: Ah, always wondered what that was for. Thanks!

File: DigitalLearningSolutions.Web.IntegrationTests/CustomWebApplicationFactory.cs (hunk 25 above)
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web.IntegrationTests/CustomWebApplicationFactory.cs:26]: Ah nice.

File: DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web/Controllers/Signposting/LinkLearningHubSsoController.cs]: Okay, sounds good.

**Review by SteveJacksonSoft — COMMENTED** [review]
Cool, think it's just the two already-existing threads to resolve.

File: DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml (hunk 12 above)
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml:57]: @SteveJacksonSoft based on @daniel-manta-hee demo today, the header in the above screenshot actually looks like it closer matches what Daniels demo presented rather than the wireframe attached to 681.
Would it then make sense to revert this change and by reverting the change we will get a result that matches the above screenshot and be inline with the other pages.

[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml:57]: @wilkinsm Following conversation with Kevin, let's just use the standard learning portal menu here (with menu as usual)

File: DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web.IntegrationTests/AuthenticationWebApplicationFactory.cs:27]: Done.

File: DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml (hunk 12 above)
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml]: I've reversed the layout changes so we are now left with the standard layout as per below:

image

[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml]: Why does the menu only have `Help` on it?

File: DigitalLearningSolutions.Web/Controllers/Signposting/SignpostingSsoController.cs [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+    using DigitalLearningSolutions.Web.ViewModels.SignpostingSso;
+    using Microsoft.AspNetCore.Authorization;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.FeatureManagement.Mvc;
+
+    [FeatureGate(FeatureFlags.UseSignposting)]
+    [Route("Signposting")]
+    [Authorize(Policy = CustomPolicies.UserOnly)]
+    public class SignpostingSsoController : Controller
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web/Controllers/Signposting/SignpostingSsoController.cs:16]: Ah, think we just want to add an attribute to say this is in the LearningPortal DlsSubApplication

File: DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml (hunk 12 above)
[the remark this answers is not in the document]
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Views/Shared/_Layout.cshtml]: Sorted see other comment.

File: DigitalLearningSolutions.Web/Controllers/Signposting/SignpostingSsoController.cs (hunk 27 above)
Comment (wilkinsm) [reply DigitalLearningSolutions.Web/Controllers/Signposting/SignpostingSsoController.cs:19]: Ah yes, done that now and it looks like this:

image

Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web/Controllers/Signposting/SignpostingSsoController.cs:19]: Beautiful!

**Review by SteveJacksonSoft — APPROVED** [review]
